Add a `[lints]` table to `Cargo.toml` (#3389)
rust-lang/rfcs: text/3389-manifest-lint.md

workspace = true
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

When parsing a manifest, cargo will resolve workspace inheritance for
`lints.workspace = true` as it does with other fields.

When running rustc, cargo will transform the lints from `lint = level` to
`--level lint` and pass them on the command line before `RUSTFLAGS`, allowing
user configuration to override package configuration.  These flags will be
finterprinted so changing them will cause a rebuild.

**Note:** This reserves the lint name `workspace` to allow workspace inheritance.

# Drawbacks
[drawbacks]: #drawbacks

A concern brought up in
[rust-lang/rust-clippy#1313](https://github.com/rust-lang/rust-clippy/issues/1313)
was that this will pass lints unconditionally to the underlying tool, leading
to "undefined lint" warnings when used on earlier versions, requiring that
warning to also be suppressed, reducing its value.  However, in the "Future
possibility's", we mention direct support for tying lints to rust versions.

This does not allow sharing lints across workspaces.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

> This could be left to `clippy.toml` but that leaves `rustc` without a solution.

Reviewer: The relationship between `clippy.toml` and `[lints]` table is not really clear here. Why do we need a new `[lints]` table given that we have `clippy.toml` today? Does `clippy.toml` become obsolete after this landed? How do we teach users to migrate one to the other, or the difference between these two files?

Sorry I don't have a constructive thought at this moment. Just try dumping things from my head.

Reviewer 2: It is not possible to set lint levels in `clippy.toml`, which is a big motivation for this RFC. We don't want to add this to `clippy.toml`, because it is way better placed in `cargo`. If you want all the details about this topic, see this issue: https://github.com/rust-lang/rust-clippy/issues/1313

Author: The main deficiency with `clippy.toml` today is that you can't override the level for a lint.

The main deficiency in the long term for `clippy.toml` for this is that it is clippy specific but we need to also deal with rustc lints and rustdoc lints now and cargo and cargo-semver-check lints in the future.

An unknown for me for `clippy.toml` is its workspace interactions.

That said, I think this is a great conversation to have with the clippy team on what their ideal future state is and which better fits that ideal future state.

Reviewer 2: From the Clippy side: We always treated `clippy.toml` as an unstable feature with the potential to be removed in the future.

With this RFC and the possibility to configure lints in `Cargo.toml` there is no more reason for `clippy.toml` to exist and we can deprecate/remove it.

So in a perfect world, lint level definition and lint configuration will be handled by `cargo` and `clippy.toml` doesn't exist.

Author: @flip1995 if this is accepted, what does the clippy team see as the role of `clippy.toml` in the long term?   If it is to be merged in to `Cargo.toml`, are there any deficiencies in this design for that, like clippy-wide configuration?  

Are there any challenges in reading configuration from `Cargo.toml` that would would need to resolve?  For example, would we need to include it in `cargo metadata` so we can process workspace inheritance for you?

Reviewer: Thanks for those pointers and information! I feel like a summary of rust-lang/rust-clippy#1313 is worth mentioning somewhere in this RFC, if it doesn't burden the RFC and its author.

Reviewer 2: `clippy.toml` allows lint configuration independent from lints. This is important because some configurations (most importantly `msrv`) is used by multiple lints. So in order to fully include the `clippy.toml` functionality in `Cargo.toml`, we'd need a way to specify keys that are **not** lints and then pass them to Clippy (or other tools).

We use `cargo metadata` (or better the cargo-metadata crate) to get information about the `Cargo.toml` file. So it would have to be included in the cargo metadata.

But I would leave this open for future discussions. For lint configurations, I'd suggest to add a `[lints.configuration]`/`[lints.<tool>.configuration]` subsection

Reviewer 2: > I feel like a summary of [rust-lang/rust-clippy#1313](https://github.com/rust-lang/rust-clippy/issues/1313) is worth mentioning somewhere in this RFC

The summary is that `clippy.toml` can't define lint levels and will not do so in the future. The rest of the issue is patiently explaining this decision to people who kept asking (which was and is a reasonable question).

Author: > So in order to fully include the clippy.toml functionality in Cargo.toml, 

Is this something the RFC should aim for though?  I want to make sure this includes the clippy team's perspective rather than being prescriptive on what clippy should do.

> we'd need a way to specify keys that are not lints and then pass them to Clippy (or other tools).

Does cargo need to pass this extra configuration or just expose it via `cargo metadata`?

If we do need to pass it, I would suggest we do it as a literal translation of command-line arguments so cargo doesn't have to hard code how to pass each one.

> But I would leave this open for future discussions. For lint configurations, I'd suggest to add a [lints.configuration]/[lints.<tool>.configuration] subsection

A part of me wants to call this "metadata" for consistency (flags it as "cargo does not care") though I feel like that could also be confusing for new users not steeped in cargo.

This also has me leaning towards `lints.<tool>.<lint>` rather than `lints.<tool>::<lint>`.

Author: > No, this is too much for this RFC IMO. But we should keep this in mind for future possibilities.

Sorry, this is what I meant.  Should the RFC be written with the expectation that clippy would want to use lint configuration in `Cargo.toml`, replacing the use of `clippy.toml`.

Author: The one question I have with `lints.<tool>.<lint>`  is whether rustc should be under
- `lints`: cleaner translation for diagnostic messages to `Cargo.toml`
- `lints.rust`: cleaner data model, more future proofed against adding new tools.

Reviewer 2: > Sorry, this is what I meant. Should the RFC be written with the expectation that clippy would want to use lint configuration in `Cargo.toml`, replacing the use of `clippy.toml`.

I would not put the main focus on this. I would focus on making defining lint levels most ergonomic. Only if it comes to 50:50 decisions I would go for the one that makes it easier to also include lint/tool configuration.

---

I would put `rustc` lints under just `[lints]`. In code Rust lints are also not scoped. So not scoping it in the Cargo.toml makes the most sense to me.

Reviewer 3: I think we should stick with the top-level table, rather than making the default case more verbose for the sake of allowing for tool lints with the same prefix as an existing rust lint. I don't think that that case would work well with the compiler either.

Author: *https://github.com/rust-lang/rfcs/pull/3389#discussion_r1107371674 is the thread about `<level> = [<lint>]` or `<lint> = <level>`*

Thinking more about this, I don't think an unscoped rust lint table is feasible with `<lint> = <level>` as we need to scrape up all of the contents and then forward them on to a tool.   Without configurable lints, this is easy as we just say "is the value a string?  Forward it" but becomes more complex with configurable lints as now we are getting into detecting the shape of a table.

That said, I do not think this immediately pushes us from one data model to the other but this is another angle for us to consider.

Author: The RFC has been updated to `lints.<tool>.<level>` with a `lints.<tool>.metadata` table in the configurable lints future possibility

Reviewer 2: I think this thread can be marked as resolved? All other topics opened up in this thread, like `<level> = [<lint>]` vs `<lint> = <level>` are addressed in different threads.

`[lints]` could be `[package.lints]`, tying it to the package unlike `[patch]`
and other fields that are more workspace related.  Instead, we used
`[dependencies]` as our model.

`[lints]` could be `[lint]` but we decided to follow the precedence of `[dependencies]`.

Instead of using `::` as a separator between tool and lint (e.g.
`clippy::enum_glob_use`), we could use TOML dotted keys for this (e.g.

> `clippy.enum_glob_use`).  This has the advantage of allowing unquoted keys at

Reviewer 2: Another possibility would be to put the tool as a sub-section of `[lints]`. So `rustc` lints  get configured in `[lints]`, Clippy lints in `[lints.clippy]` and from other tools in `[lints.<tool-name>]`.

Author: The data model between what you mentioned and what is in the RFC is the same, just using two different ways of expressing it within TOML syntax.

Reviewer 2: Ah right, then I consider this a really good idea! :)

the cost of not being able to copy/paste the lint name from the tool's output

> into the fileV

Reviewer 4: ```suggestion
into the file.
```

- We have not yet defined semantics for sharing something like this across a
  workspace

Instead of using workspace inheritance for `[lint]`, we could make it
workspace-level configuration, like `[patch]` which is automatically applied to
all workspace members.  However, `[patch]` and friends are because they affect
the resolver / `Cargo.toml` and so they can only operate at the workspace
level.  `[lints]` is more like `[dependencies]` in being something that applies
at the package level but we want shared across workspaces.

Instead of traditional workspace inheritance where there is a single value to
inherit with `workspace = true`, we could have `[workspace.lints.<preset>]`
which defines presets and the user could do `lints.<preset> = true`.  The user
could then name them as they wish to avoid collision with rustc lints.

Instead of the `[lints]` table being `lint = "level"`, we could organize
it around `level = ["lint", ...]` like some other linters do (like
[ruff](https://beta.ruff.rs/docs/configuration/)) but this works better for
logically organizing lints, highlighting what changed in diffs, and for
possibly adding lint-specific configuration in the future.

# Prior art

> [prior-art]: #prior-art

Reviewer 2: There is also [`cargo cranky`](https://github.com/ericseppanen/cargo-cranky) that is worth mentioning here. This also uses the `level = [ ... ]` syntax.

Author: Good call; now included

Javascript
- [eslint](https://eslint.org/docs/latest/use/configure/rules)
  - Format is `lint = level` / `lint = [ level, additional config ]`

Go
- [golangci-lint](https://golangci-lint.run/usage/configuration/)
  - Format is `level = [lint, ...]`

Ruby
- [rubocop](https://docs.rubocop.org/rubocop/1.45/configuration.html)
  - Format is `Lint: Enabled: true`

# Unresolved questions
[unresolved-questions]: #unresolved-questions

How should we hand rustdoc lint levels or, in the future, cargo lint levels?
The current proposal takes all lints and passes them to rustc like `RUSTFLAGS`
but rustdoc uses `RUSTDOCFLAGS` and cargo would use neither.  This also starts
to get into
[user-defined tool attributes](https://rust-lang.github.io/rfcs/2103-tool-attributes.html).

Should we only apply/fingerprint lints for the appropriate tool?  For example,
we would not include and fingerprint `clippy::` lints when running builds,
allowing them to change without forcing a rebuild.  We likely already need to
be tool-aware for built-in tools to handle `rustdoc::` lints (see above) so
this isn't much more of a step.

> How do we allow controling precedence between lints and lint groups?  We are

Reviewer 5: I'm imagining these are treated as rustc flags, right? There's already a precedence between flags / code and groups

Author: Isn't the precedence based on the order they appear on the command line?

If so, when we translate the fields
```toml
future-incompatible = "deny"
semicolon_in_expressions_from_macros = "allow"
```
to rustc flags, how do we know it should be translated to `-D future-incompatible -A semicolon_in_expressions_from_macros`?

While to the user, the TOML appears ordered, by definition it isn't and it is dependent on whether we use `indexmap`, `BTreeMap`, or `HashMap` for what ordering we get and that would be an implementation detail bleeding out.

Reviewer 5: > Isn't the precedence based on the order they appear on the command line?

Oh, for groups vs levels, yes

Perhaps we can specify it as groups-first-then-individuals?

Author: How would cargo know what is a group?  We could hard code the list and try to keep it up-to-date but that makes third-party tool lints even more difficult.

Author: With our [larger thread on precedence](https://github.com/rust-lang/rfcs/pull/3389#discussion_r1133265436), I'm marking this as resolved to encourage the conversation over there

using a TOML table with the keys as lint names which does not allow controlling

> ordering.  Even if we switched to `level = [lint, ...]`, you get a hard coded

Reviewer 2: I think with `level = [lint, ...]` this is an edge case: Only if a user would list a lint in two different levels, we would run into this. And for this we could define the precedence to always pass `allow` flags first, then `warn` and so on.

Author: Its not passing a lint at two different levels but a group and a lint.  

Also, that hard coded precedence doesn't help in the example ehuss brought this up in: `-D future-incompatible -A semicolon_in_expressions_from_macros`

Reviewer 2: Ah I see, I missed the group part, sorry.

I think it would make the most sense to pass groups first and single lints second. But this leaves open the questions how to differentiate between groups and lints :thinking: I don't think there exists anything for that today.

Reviewer 2: Marking this as resolved in favor of this thread: https://github.com/rust-lang/rfcs/pull/3389#discussion_r1107358269

precedence between levels that the user can't control.

> # Future possibilities

Reviewer 2: One future possibility I'd like to see here would be an option to link to a file that defines lint levels. A lot of people would prefer an extra file (me included) that can just be copied across projects. Instead of always copying the list out of the `Cargo.toml`, people would then just have to mention the file in the `Cargo.toml` and can copy it around (or submodule it from a central repo).

This would be a good compromise of putting the configuration inside `Cargo.toml` instead of a `lints.toml` file.

I would keep this as a future possibility, because I don't consider this important for the MVP.

Reviewer 5: Yeah, that would be nice.

Author: I overlooked that because I already have to diff `Cargo.toml` files between all my projects as I percolate out changes in my practices.

I've now added it to future possibilities, pointing to `package.license` as precedence.

[future-possibilities]: #future-possibilities

## Configurable lints

We can extend basic lint syntax:
```toml
[lints]
cyclomatic_complexity = "allow"
```
to support configuration, whether for cargo or the lint tool:
```toml
[lints]

> cyclomatic_complexity = { level = "allow", rust-version = "1.23.0", threshold = 30 }

Reviewer 6: This syntax is inverted compared to `#[allow(lint1, lint2)]`, which surprised me.

I find the form below clearer and it encourages clearly separating lints by level

```toml
[lints]
allow = [
    "unsafe",
    { name = "cyclomatic_complexity", rust-version = "1.23.0", threshold = 30 },
]
warn = [
    # ...
]
```

It's probably closer to how `(dev-)dependencies` work too:  we don't specify whether the dependency is dev, build or normal in `clap = { ... }`, we specify it outside

Ultimately I think both are fine, I just prefer the above form with lists.

The downside is that changing a lint from one level to another is now a bigger diff and less "local"

Author: This was covered down in the Alternatives section

> Instead of the `[lints]` table being `lint = "level"`, we could organize
it around `level = ["lint", ...]` like some other linters do (like
[ruff](https://beta.ruff.rs/docs/configuration/)) but this works better for
logically organizing lints, highlighting what changed in diffs, and for
possibly adding lint-specific configuration in the future.

Yes, technically, we can put inline tables into inline arrays, like your example, but it is getting a bit messier.

Reviewer 2: I agree that this feels way more natural. Also with allowing `{ ... }` instead of just lint names in the list, we also get configurability here. However, this would necessitate to define a lint level in order to configure lints, which is probably unnecessary in most cases.

Reviewer 6: > This was covered down in the Alternatives section

Ah sorry, missed it. 

> I forgot to include in that quoted block that you also have to define the semantics for when a lint appears in multiple levels.

An error from cargo should be enough ?

---

> works better for logically organizing lints, highlighting what changed in diffs, and for possibly adding lint-specific configuration in the future.

I don't think that's true apart from diffs

In all projects where I have a `[target.*.rustflags]`, the lints are separated by allow/warn/deny and then sorted if necessary. In `clippy.toml` the configuration are sorted per-lint but there is no notion of level there.

> lint-specific configuration in the future.

How does the following work in toml ? Does one overrides the other ?

```toml
[lints.allow."unsafe"]
config1 = true

[lints]
allow = ["lint-2", "lint-3"]
```

Author: > How does the following work in toml ? Does one overrides the other ?

Assigning an array to an implicit table is a hard error.

Reviewer 6: ah, there goes my hope, then you solution is probably best

```
Where `rust-version` is used by cargo to determine whether to pass along this
lint and `threshold` is used by the tool.  We'd need to define how to
distinguish between reserved and unreserved field names.

> ## Extending the syntax to `.cargo/config.toml`

Reviewer 5: I actually think this is probably necessary for this RFC to make sense. I feel this makes more sense in the local (potentially-but-not-necessarily checked-in) config as opposed to the crate config.

Reviewer 5: Though with workspace inheritance perhaps it's not that necessary.

Author: Could you clarify why you feel it makes more sense to put this in `.cargo/config.toml` over `Cargo.toml`?

In starting this with `Cargo.toml`, I'm looking to address the existing problems with doing this via rustflags in `.cargo/config.toml` (project configuration being directory dependent).  https://github.com/EmbarkStudios/rust-ecosystem/issues/59 is a good example of people using both source code and `.cargo/config.toml` for lints that are project-specific and would benefit from this.   This was also spurred by the cargo team itself wanting to use this feature.

Reviewer 5: I think it makes sense to allow it in _both_.

The reason being that there are multiple lint workflows: sometimes you lint in CI and want everyone to follow the same lint set. However, some people may wish to opt in to additional lints locally for their own code, and being able to set that is useful.

But I'm fine with starting out in Cargo.toml at least. I would _prefer_ to try and solve it for both.

Author: > The reason being that there are multiple lint workflows: sometimes you lint in CI and want everyone to follow the same lint set. However, some people may wish to opt in to additional lints locally for their own code, and being able to set that is useful.

The part I'm missing is why users can't just use rustflags for this in the short term?  In the long term, if we add configurable lints, there isn't an alternative and I can see it becoming a higher priority then.  This is the main reason I put it as a future possibility, rather than including it.

Reviewer 3: I don't think we should add this to config.toml.

People can still do this via rustflags if they really want to, but I don't think we should encourage a workflow in which some lints are enabled in some build contexts but not in others.

Reviewer 7: > @joshtriplett I don't think we should encourage a workflow in which some lints are enabled in some build contexts but not in others.

So you're saying that I shouldn't be able to easily enable custom lints on my machine if the projects don't already use them?

There are many Clippy lints which have high false positive rates, too restrictive or even self-contradictory. `pedantic` and `restriction` categories are full of those. They are infeasible to gate CI on, which means they shouldn't be enabled in the project's source or `Cargo.toml`. But there is no reason to not allow easily running those locally. Neither should we make it hard to locally disable any lints, including all warnings. The quality control is enforced by CI anyway, and if I'm digging in the Git history, lint warnings are just syntactic noise which can't be acted on anyway.

The rustflags solution specifically is bad, because setting it blows the build cache. Again, if I'm digging Git history, or checking out someone else's branch, or just taking a look at some random project on Github, I'd usually want custom lint settings (sometimes more lax than the defaults, sometimes more strict). Setting rustflags would mean that all my cached dependencies would need to be recompiled, for no good reason. It breaks incremental compilation, increases compile times, increases power usage and SSD wear, and all for no reason other than Cargo maintainers' personal taste. I could easily set those for rustc if I used a different build system, the compiler already supports it, it's just that Cargo doesn't want to make it easy to use.

Author: > The rustflags solution specifically is bad, because setting it blows the build cache. Again, if I'm digging Git history, or checking out someone else's branch, or just taking a look at some random project on Github, I'd usually want custom lint settings (sometimes more lax than the defaults, sometimes more strict). Setting rustflags would mean that all my cached dependencies would need to be recompiled, for no good reason. 

This wasn't too clear to me.  We will still be fingerprinting `[lints]` and causing rebuilds, so having support for `[lints]` in `.cargo/config.toml` will be no different for rebuilds than rustflags.  Well, we might specialize the fingerprinting for the tool (e.g. don't fingerprint clippy lints for rustc builds).  Is that the part that you feel would be an improvement?

Reviewer 8: When you are cross-compiling, there is no way to set rustflags via `.config/cargo.toml` that applies *to everything* (built for host or built for target), so even just for having something that works across proc-macros and normal crates within a workspace, this would be great.

Reviewer 7: > We will still be fingerprinting [lints] and causing rebuilds

Why? Lints don't affect the compiled binaries. It seems like pure overhead to me.

----- 

While I'd want to be able to set lints globally (for every project on my computer), there is a design issue of config priority which is unclear to me. Currently afaik the only option duplicated in both `config.toml` and `Cargo.toml` is [`[profile]`](https://doc.rust-lang.org/cargo/reference/config.html#profile), and in its case the `config.toml` has priority. In particular, a `[profile]` section in `$HOME/.cargo/config.toml` would override the `[profile]` in any other Cargo project.

That may be reasonable for `[profile]`, but would likely be incorrect for lints. Most of the time I'd want to use the lints from the specific project I work on, with the `config.toml` defaults (particularly defined in non-project directories) used only occasionally. So the correct priority looks like `$WORKSPACE/.cargo/config.toml` > `$WORKSPACE/subcrate/Cargo.toml` > `$WORKSPACE/Cargo.toml` > `$UPPER_DIR/.cargo/config.toml`. This looks a bit convoluted and inconsistent with the behaviour of other `config.toml` keys.

Maybe there should be a cargo command-line flag which would control whether `Cargo.toml` or `config.toml` lint settings should be used?

Actually, if `cargo build/run/test` would have flags to set lint levels per invocation, that would be sufficient. That would also avoid the question of config priorities.

Author: > Why? Lints don't affect the compiled binaries. It seems like pure overhead to me.

If a user adds additional lints and run `cargo check`, they would expect to see the result of those lints.  This would only be possible by rebuilding which will require finger printing.

> While I'd want to be able to set lints globally (for every project on my computer), there is a design issue of config priority which is unclear to me. 

For now, I am leaving `[lints]` for `.cargo/config.toml` as a future possibility rather than removing it or making it required to move this proposal along.     Keeping it as a future possibility has no cost.  Adding it into the proposal has a larger cost, in design, decision making, and implementation.  It would require a very strong motivation to block incremental improvement (`[lints]` in `Cargo.toml`) on getting a full solution (e.g. including `[lints]` in `.cargo/config.toml`).

With that context, maybe we can temper how much design work we put into the feature or how much we debate whether we should or shouldn't do it in the future.

Reviewer 7: Sure, I wouldn't want future extensions to block this long-waited feature either. 

When you say "rebuilding", do you mean rebuilding the workspace crates or all dependencies? I am most concerned about the latter, even though for large projects there could be no difference. I am not familiar with the inner details of `cargo check`.

*as a new ["Manifest Format" entry](https://doc.rust-lang.org/cargo/reference/manifest.html#the-manifest-format)*

Override the default level of lints by assigning them to a new level in a
table, for example:
```toml
[lints]
unsafe = "forbid"
```

Supported levels include:
- `forbid`
- `deny`
- `warn`
- `allow`

**Note:** TOML does not support `:` in unquoted keys, requiring tool-specific
lints to be quoted, like
```toml
[lints]
"clippy::enum_glob_use" = "warn"
```

## The `lints` table

*as a new [`[workspace]` entry](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-workspace-section)*

The `workspace.lints` table is where you define lint configuration to be inherited by members of a workspace.

Specifying a workspace lint configuration is similar to package lints.

Example:

```toml
# [PROJECT_DIR]/Cargo.toml
[workspace]
members = ["crates/*"]

[workspace.lints]
unsafe = "forbid"
```

```toml
# [PROJECT_DIR]/crates/bar/Cargo.toml
[package]
name = "bar"
version = "0.1.0"

[lints]
workspace = true
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

When parsing a manifest, cargo will resolve workspace inheritance for

> `lints.workspace = true` as it does with other fields.

Reviewer 8: This means that I can write `lints = { workspace = true, foo = "deny" }` in a specific crate within a workspace to inherit the lint levels defined in the workspace manifest, but overwrite the lint level for `foo`, right?

Author: I made this more explicit in that I intended this to error but I put merging in as a future possibility.

The current proposal takes all lints and passes them to rustc like `RUSTFLAGS`
but rustdoc uses `RUSTDOCFLAGS` and cargo would use neither.  This also starts
to get into
[user-defined tool attributes](https://rust-lang.github.io/rfcs/2103-tool-attributes.html).

Should we only apply/fingerprint lints for the appropriate tool?  For example,
we would not include and fingerprint `clippy::` lints when running builds,
allowing them to change without forcing a rebuild.  We likely already need to
be tool-aware for built-in tools to handle `rustdoc::` lints (see above) so
this isn't much more of a step.

How do we allow controlling precedence between lints and lint groups?  We are
using a TOML table with the keys as lint names which does not allow controlling
ordering.  Even if we switched to `level = [lint, ...]`, you get a hard coded
precedence between levels that the user can't control.

# Future possibilities
[future-possibilities]: #future-possibilities

## Configurable lints

We can extend basic lint syntax:
```toml
[lints]
cyclomatic_complexity = "allow"
```
to support configuration, whether for cargo or the lint tool:
```toml
[lints]

> cyclomatic_complexity = { level = "allow", rust-version = "1.23.0", threshold = 30 }

Reviewer 2: Something else to consider would be to allow both: Defining levels with `level = ["lint", ...]` AND (`lint = "level"` OR `lint = { ... }`).

Author: Should we pull this into the other thread about the data model?

Reviewer 2: Sure. I extracted it from this discussion, because I rather see this as a future addition, once lint configuration is added. So in the MVP I would go for _either_ `lint = "level"` OR `level = ["lint", ...]` and then extend this once we need it for lint configurations. (I prefer the `level = ["lint", ...]` way for the MVP).

Author: How does lint configuration change it?

Personally, I think we should only have one syntax so we should consider the likelihood of the future state and plan to that now.

Reviewer 2: Yeah, thinking about this more, I think only one syntax makes more sense. 

Especially because lint configuration (from a Clippy perspective) is not about configuring a lint with a value, but configuring the tool, which will then adapt the behavior of one or many lints depending on the configuration value. So I think lint/tool configuration should be separated from the decision of the format.

The only lint-specific configuration I can see are configurations of the lint levels and how cargo should treat them, i.e. `rust-version` configuration attached to the level. IMO this is best achieved with this suggestion: https://github.com/rust-lang/rfcs/pull/3389#discussion_r1107371674

Author: > Especially because lint configuration (from a Clippy perspective) is not about configuring a lint with a value, but configuring the tool, which will then adapt the behavior of one or many lints depending on the configuration value

From my perspective as a user, some of the configuration is general, like `avoid-breaking-exported-api`, while others are lint specific, like `cognitive-complexity-threshold`. 

Is there a reason this model is inherently incorrect that we couldn't model it this way in the tool?  For example, past discussions of lint configuration have assumed `cyclomatic_complexity = { state = "deny", threshold = 30 }` would the way to do it, rather than setting the level for `cyclomatic_complexity` in one place and then setting the threshold in another.

Reviewer 2: > Is there a reason this model is inherently incorrect that we couldn't model it this way in the tool?

Lint configuration in Clippy is always 1-to-many. For some/most configurations it is indeed 1-to-1. But this could just change in the future and another lint could just also pick up this configuration value and use it. 

Only being able to configure one lint at a time would make things worse IMO. E.g. for the `msrv` config you would have to add >15 lines of lints with the same `msrv` config applied.

So I would stop calling this _lint_ configuration, because _lint_ configuration doesn't really exist from a Clippy perspective. I would call it _tool_ configuration, which in the case of Clippy configures lints. And with that I wouldn't try to cover this in this RFC. Not even in a `[lints.<tool>.metadata]` section.

TL;DR: The ~~cake~~ lint configuration is a lie :cake:

into the file.

We could support platform or feature specific settings, like with
`[lints.<target>]` or `[target.<target>.lints]` but
- There isn't a defined use case for this yet besides having support for `cfg(feature = "clippy")` or
  which does not seem high enough priority to design
  around.
- `[lints.<target>]` runs into ambiguity issues around what is a `<target>`
  entry vs a `<lint>` entry in the `[lints]` table.
- We have not yet defined semantics for sharing something like this across a
  workspace

Instead of using workspace inheritance for `[lint]`, we could make it
workspace-level configuration, like `[patch]` which is automatically applied to
all workspace members.  However, `[patch]` and friends are because they affect
the resolver / `Cargo.toml` and so they can only operate at the workspace
level.  `[lints]` is more like `[dependencies]` in being something that applies
at the package level but we want shared across workspaces.

Instead of traditional workspace inheritance where there is a single value to
inherit with `workspace = true`, we could have `[workspace.lints.<preset>]`
which defines presets and the user could do `lints.<preset> = true`.  The user

> could then name them as they wish to avoid collision with rustc lints.

Author: I wanted to call attention to this item.  Would presets or custom lint groups of interest enough for us to design around?

Python
- [flake8](https://flake8.pycqa.org/en/latest/user/configuration.html)
  - Format is `level = [lint, ...]`
- [pylint](https://github.com/PyCQA/pylint/blob/main/examples/pylintrc#L402)
  - Format is `level = [lint, ...]` but the [config file is a reflection of the CLI](https://pylint.pycqa.org/en/latest/user_guide/configuration/index.html)
- [ruff](https://beta.ruff.rs/docs/configuration/)
  - Format is `level = [lint, ...]`, likely due to past precedence in ecosystem (see above)

Javascript
- [eslint](https://eslint.org/docs/latest/use/configure/rules)
  - Format is `lint = level` / `lint = [ level, additional config ]`

Go
- [golangci-lint](https://golangci-lint.run/usage/configuration/)
  - Format is `level = [lint, ...]`

Ruby
- [rubocop](https://docs.rubocop.org/rubocop/1.45/configuration.html)
  - Format is `Lint: Enabled: true`

# Unresolved questions
[unresolved-questions]: #unresolved-questions

How should we hand rustdoc lint levels or, in the future, cargo lint levels?
The current proposal takes all lints and passes them to rustc like `RUSTFLAGS`
but rustdoc uses `RUSTDOCFLAGS` and cargo would use neither.  This also starts
to get into

> [user-defined tool attributes](https://rust-lang.github.io/rfcs/2103-tool-attributes.html).

Author: Would it work for cargo to just hard code what lint tool goes to what CLI? 

The problem is then dealing with the potential future state of third-party lint tools.  If those were opted into via cargo instead of a source-level attribute, we might be able to get the information we need to direct the lints to the appropriate CLI (or ignore as needed)

Reviewer 2: Couldn't cargo always pass all lints defined in the `[lint]` section to all tools? If a tool is not used and therefore not registered with `rustc`, `rustc` will just ignore that lint and that it was passed to it. Or am I misunderstanding the problem? :thinking:

Author: I've not touched that side of things so I am unsure if that would be a problem or not.

ehuss also brought up the possibility of using this to minimize which lints are included in fingerprinting for reducing recompiles for lint changes (e.g. adding clippy lints doesn't cause `cargo build` to rebuild).  Its mentioned in the RFC just after this paragraph.

A concern brought up in
[rust-lang/rust-clippy#1313](https://github.com/rust-lang/rust-clippy/issues/1313)
was that this will pass lints unconditionally to the underlying tool, leading
to "undefined lint" warnings when used on earlier versions, requiring that
warning to also be suppressed, reducing its value.  However, in the "Future
possibility's", we mention direct support for tying lints to rust versions.

This does not allow sharing lints across workspaces.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

This could be left to `clippy.toml` but that leaves `rustc` without a solution.

`[lints]` could be `[package.lints]`, tying it to the package unlike `[patch]`
and other fields that are more workspace related.  Instead, we used
`[dependencies]` as our model.

`[lints]` could be `[lint]` but we decided to follow the precedence of `[dependencies]`.

Instead of using `::` as a separator between tool and lint (e.g.
`clippy::enum_glob_use`), we could use TOML dotted keys for this (e.g.
`clippy.enum_glob_use`).  This has the advantage of allowing unquoted keys at
the cost of not being able to copy/paste the lint name from the tool's output

> into the file.

Reviewer 3: I think we should do this; the subtable model seems like a perfect fit for tool lints, and the syntax feels much easier to work with without having to quote the names (which already makes it not a clean copy and paste).

Author: RFC has been updated.  I was originally waiting to see where other threads went but figured I'd take a stab at it now and I can change it later.

The current proposal takes all lints and passes them to rustc like `RUSTFLAGS`
but rustdoc uses `RUSTDOCFLAGS` and cargo would use neither.  This also starts
to get into
[user-defined tool attributes](https://rust-lang.github.io/rfcs/2103-tool-attributes.html).

Should we only apply/fingerprint lints for the appropriate tool?  For example,
we would not include and fingerprint `clippy::` lints when running builds,
allowing them to change without forcing a rebuild.  We likely already need to
be tool-aware for built-in tools to handle `rustdoc::` lints (see above) so
this isn't much more of a step.

How do we allow controlling precedence between lints and lint groups?  We are
using a TOML table with the keys as lint names which does not allow controlling
ordering.  Even if we switched to `level = [lint, ...]`, you get a hard coded
precedence between levels that the user can't control.

What does this mean for rustfmt?  If this feature slowly absorbs the role of
`clippy.toml`, a `rustfmt.toml` file might stick out like a sore thumb.  Do we
accept that, shoe-horn it in, generalize this feature into "tools" and "rules"

> from "linters" and "lints"?

Author: @CobaltCause 

> > should we generalize the RFC to be about "tools" and "rules", rather than "linters" and "lints"?
> 
> I'm inclined to think so. To me, "linters" and "lints" seems like a special case of "tools" and "configuration" (not totally sure what's meant by "rules"). On this topic, I made [this comment](https://github.com/rust-lang/cargo/issues/5034#issuecomment-1244722861) on the cargo issue, pointing out [Python's `pyproject.toml` `[tool]` section from PEP 518](https://peps.python.org/pep-0518/#tool-table) as prior art for this idea. I think tools like [`cargo-audit`](https://github.com/rustsec/rustsec/tree/main/cargo-audit) and [`cargo-embed`](https://github.com/probe-rs/probe-rs/tree/master/cargo-embed) would benefit from a general solution like that (if not from `workspace-deduplicate`, at least from not having a bunch of config files laying around).

Author: There was a [prior proposal for a `tool` section](https://internals.rust-lang.org/t/tool-configs-in-cargo-toml-e-g-rustfmt-clippy-and-alternatives-brainstorming/9055).

In the linked issue, you mentioned

> [Cargo already has [package.metadata] and [workspace.metadata]](https://doc.rust-lang.org/cargo/reference/manifest.html#the-metadata-table) too which sound pretty similar.

For me, the question is what is the value add of generalizing `lint` to `tool` in light of us having `metadata`?

I think the value of `lint` is that it is more structured, closed-ended design where cargo can more specifically act on it.

Reviewer 9: Perhaps there's a third option where you have `[lint]` but it is only for rustc lints. Clippy lint config can be under `[tool.clippy.lint]`. That is, each tool can define its own lint config (that should follow some conventions).

Reviewer 10: > There was a [prior proposal for a `tool` section](https://internals.rust-lang.org/t/tool-configs-in-cargo-toml-e-g-rustfmt-clippy-and-alternatives-brainstorming/9055).

Interesting, I hadn't seen that proposal before; the author articulates their reasoning a lot better than I could have :sweat_smile:

> For me, the question is what is the value add of generalizing `lint` to `tool` in light of us having `metadata`?

Reading through [the documentation for `{package,workspace}.metadata`](https://doc.rust-lang.org/cargo/reference/manifest.html#the-metadata-table) and [the `workspace-deduplicate` RFC](https://rust-lang.github.io/rfcs/2906-cargo-workspace-deduplicate.html?highlight=2906#placeholder-values), I think the technical value-add is that `{workspace,package}.metadata` don't support workspace inheritance whereas this RFC does. There's also the social value-add of this RFC, which is that it uses stronger language (like "is" instead of "can"), which might lead to wider adoption (assuming a reason we don't see `{workspace,package}.metadata` being used broadly is due to uncertainty in tool developers about whether they should use those).

> I think the value of `lint` is that it is more structured, closed-ended design where cargo can more specifically act on it.

I guess I'm missing something because this makes it sound like some functionality would be lost if the current state of this RFC were generalized; is that the case?

Author: > Reading through [the documentation for {package,workspace}.metadata](https://doc.rust-lang.org/cargo/reference/manifest.html#the-metadata-table) and [the workspace-deduplicate RFC](https://rust-lang.github.io/rfcs/2906-cargo-workspace-deduplicate.html?highlight=2906#placeholder-values), I think the technical value-add is that {workspace,package}.metadata don't support workspace inheritance whereas this RFC does. T

The more open-ended we make this, the less flexibility we have.  For example, a challenge with doing workspace inheritance with `metadata` is that its design is left open ended and we might start interpreting things as they weren't intended.  Also, one of the future possibilities in this RFC is to extend workspace inheritance so the two levels can be merged.  It would be difficult to support that with an open ended design as we don't know enough about the data to correctly merge it.

> (assuming a reason we don't see {workspace,package}.metadata being used broadly is due to uncertainty in tool developers about whether they should use those).

I think we should ask them, rather than assume why they don't use it.

Also as a counter example, I do support it in `cargo release`.  I do implicit workspace inheritance though as I supported that before the workspace inheritance feature was implemented.

> I guess I'm missing something because this makes it sound like some functionality would be lost if the current state of this RFC were generalized; is that the case?

I think we could generalize the shape of the data but if we generalize it too far, then this just becomes another `metadata` table.

Reviewer 11: I will also speak into the rustfmt specifics of this, but I first want to underscore my view that it's important to consider this holistically around the experience and expectations developers may have for configuration across the various official tools if we move ahead with this proposal.

I do not have the receipts handy, but I'm sure many have seen/heard comments to the effect of "I have `edition` set to `2018` in Cargo.toml, but when I run `rustc ...` it's using edition 2015 instead of edition 2018".

Essentially, it's a relatively common occurrence for some developers to (incorrectly) assume that info defined in Cargo.toml is honored by all tools even outside a cargo context. While this isn't necessarily a problem for tools typically, or always, invoked as part of some `cargo ...` subcommand flow, it's a persistent topic that pops up for tools that are utilized from both cargo and non-cargo contexts (`rustc`, `rustdoc`, `rustfmt`).

Fundamentally, my concern is that supporting more configuration values in Cargo.toml could exacerbate this issue. If this proceeds, then I'd love to find a way to help ameliorate this by somehow increasing awareness of when values in Cargo.toml can be used/will be honored, and crucially, when they won't. (though unfortunately I've no prescriptions).

<hr />

As for rustfmt, quick overview for anyone unfamiliar

`rustfmt` can and often is invoked in a cargo context via `cargo fmt`, but it's also commonly invoked directly both by humans and machines (e.g. editor/ide use cases to format single file, subset of a single file, etc.). In the `cargo fmt` scenario obviously relevant Cargo.toml info is utilized, but `rustfmt ...` itself doesn't, so when running rustfmt directly, one needs to explicitly specify the edition to use a non-2015 edition. rustfmt also supports various [config options](https://rust-lang.github.io/rustfmt/?version=v1.5.1&search=) that can be specified in a `rustfmt.toml` file.

We've had requests to support allowing rustfmt config options to also be defined in Cargo.toml, but it's something we've always pushed back on for a number of reasons, one of which being that `rustfmt` would either have to ignore the edition value in Cargo.toml while honoring other aspects of Cargo.toml, or essentially introduce breaking behavior where the default/implicit edition rustfmt would utilize would change in some contexts (since the ecosystem default is still 2015 and rustfmt does not currently utilize Cargo.toml defined edition, like rustc and rustdoc)

<hr />

I understand this proposal is focused on lints, and it's not my intent to scope creep. I would, however, feel more comfortable if we could either set an upper bound on how far this could extend to other tools (explicitly, even if only notionally, to not expand beyond lints), or determine if this could/should expand, and if so then try to address the potential challenges this could present for other tools

Reviewer 12: This reminds me of how "restic" (https://github.com/restic/restic) handles this. They only support command-line options (and env variables) with the rationale that config files are something that can easily be implemented by wrappers and doesn't need support by the tool.

Initially, I was annoyed by the lack of a config file but it makes sense from a maintainer perspective. Users can build wrappers with a config file format that suits their needs.

Translating this into Rust land, it would mean:

- Low-level tools like rustc, rustfmt, rustdoc, clippy etc should not support config files
- Wrapper tools like cargo can define a config file

One challenge to solve here would be how e.g. lint-specific config values are passed on the command-line.

Reviewer 10: > The more open-ended we make this, the less flexibility we have. For example, a challenge with doing workspace inheritance with `metadata` is that its design is left open ended and we might start interpreting things as they weren't intended. Also, one of the future possibilities in this RFC is to extend workspace inheritance so the two levels can be merged. It would be difficult to support that with an open ended design as we don't know enough about the data to correctly merge it.

Okay, this answers my question about the possibility of lost functionality when increasing generality, especially the point about merging configuration (the answer is "yes").

> I think we should ask them, rather than assume why they don't use it.

Good point...

> I do implicit workspace inheritance though as I supported that before the workspace inheritance feature was implemented.

I think this is a reasonable solution for workspace inheritance without real workspace inheritance, since like you said, if the structure is too general then there's no way for `cargo` to know how to correctly do the merging. Doing it this way lets each tool that wants this decide whether and how for themselves, which is more flexible.

---

> Fundamentally, my concern is that supporting more configuration values in Cargo.toml could exacerbate this issue.

I think this is already going to be a problem for `rustc` and `[lints.rust]` as defined in the current state of this RFC, so it would probably be good to "find a way to help ameliorate this by somehow increasing awareness of when values in Cargo.toml can be used/will be honored, and crucially, when they won't" either way. (I don't have any prescriptions here either.)

> `rustfmt` would either have to ignore the edition value in Cargo.toml while honoring other aspects of Cargo.toml, or essentially introduce breaking behavior where the default/implicit edition rustfmt would utilize would change in some contexts (since the ecosystem default is still 2015 and rustfmt does not currently utilize Cargo.toml defined edition, like rustc and rustdoc)

Hmm, this is unfortunate. I wonder if a boolean option like `respect_edition_from_cargo_dot_toml` in `rustfmt` that defaults to `false` (then document this as being recommended for projects that also use `cargo`) would help?

---

I've suddenly realized what the "'tools' and 'rules'" option was about, and I think this would be a good middle ground. My revised thinking is that this RFC should consider changing from `[lints]` to `[rules]` and keep pretty much everything else the same. I think `[rules]` is just enough additional generality over `[lints]` because "lint" usually means the effect is just that some warning message(s) will be generated at some point, whereas a "rule" may actually be acted upon in some general way, for example `rustfmt` formatting a file.

This should work with workspace inheritance because this RFC already makes space for that, and it should also work with the future possibility of "extend[ing] workspace inheritance so the two levels can be merged". I'm imagining this would happen like this:

```toml
# At the workspace level
[rules.some_tool]
rule_a = "x"
rule_b = "y"
```
\+
```toml
# At the package level
[rules.some_tool]
# Maybe with a `workspace = true` here as well
rule_b = { value = "z", option = true }
rule_c = "w"
```
=
```toml
# Resultant merged rules for the package
rule_a = "x"
rule_b = { value = "z", option = true }
rule_c = "w"
```

Where the right hand side of `=` in the TOML is treated as just a blob and updated wholesale, as shown.

I think this behavior could be useful for more things than just linters, like maybe parts of `cargo-audit` and definitely `rustfmt` (in a hypothetical world where `rustfmt` wants to support this, I guess).

Author: Not going to have time for all of this right now but wanted to call this out

> I've suddenly realized what the "'tools' and 'rules'" option was about, and I think this would be a good middle ground. 

I got the name "rules" from [eslint](https://eslint.org/docs/latest/use/configure/rules).  Ironic for me to not call out the prior art in the threads as I'm barely familiar with that ecosystem (it is more generically linked to in the RFC).  As that, on a surface level, seems to be both tool-agnostic while still being linter-related, I figure that is the closest prior art to what we are looking at.  I should read up further on how its designed.  If anyone has experience with it, I'd also be interested in hearing how eslint's configuration system worked out for people.

Reviewer 13: > We've had requests to support allowing rustfmt config options to also be defined in Cargo.toml, but it's something we've always pushed back on for a number of reasons, one of which being that rustfmt would either have to ignore the edition value in Cargo.toml while honoring other aspects of Cargo.toml, or essentially introduce breaking behavior where the default/implicit edition rustfmt would utilize would change in some contexts (since the ecosystem default is still 2015 and rustfmt does not currently utilize Cargo.toml defined edition, like rustc and rustdoc)

My expectation would be that invoking `rustfmt` directly behaves the same as `rustc`/`rustdoc` and its current behavior wrt. `package.edition`, it would not read the config from `Cargo.toml` itself, `cargo fmt` would read the `Cargo.toml` and translate that into CLI options. The one inconsistency then would be the fact that `rustfmt` has a config file at all, `rustc` etc. doesn't and is only configured via CLI options and relies on wrappers like `cargo rustc` to provide the config file support.

Reviewer 11: > cargo fmt would read the Cargo.toml and translate that into CLI options.

I recognize that on paper this sounds simple and straightforward, but in actuality it could get quite messy, quite quickly. Most (if not all) editor/ide format on save behaviors run `rustfmt ...` directly so if this were to happen I fear it would be far too easy for users to find themselves where their autosave formatting in the editor would conflict a `cargo fmt` run executed manually and/or as a CI gate, which would be maddening to say the least.

It would also complicate rustfmt's existing support for project/directory level formatting configurations since CLI config args override rustfmt.toml config.

Basically, rustfmt faced and tackled this same problem quite some time ago but opted for the opposite solution (everything in `*tool*.toml` instead of `Cargo.toml`). If the Cargo.toml route is indeed selected for clippy and rustc lint configs then I just wanted to make sure that we're either explicit about this not inherently being expected to cascade out to other tools, or if we do want a more consistent experience across tools, then at least acknowledge that some of these thorny points would need to be addressed before that could happen

Reviewer 2: I think we should keep this RFC about a `[lints]` section and clearly separate it from a tools configuration section. IMO this is a whole different topic ...

> One challenge to solve here would be how e.g. lint-specific config values are passed on the command-line.

... The main reason for this is that there _aren't_ "lint-specific config values", at least not in Clippy. Some of the config values you can set in `clippy.toml` might only affect a single lint, but it still configures the tool, that then changes its behavior (for one lint), rather than the lint directly.

More details: https://github.com/rust-lang/rfcs/pull/3389#discussion_r1108211518

> Basically, rustfmt faced and tackled this same problem quite some time ago but opted for the opposite solution (everything in `*tool*.toml` instead of `Cargo.toml`)

Clippy did the same thing and has a `clippy.toml` for configuring the tool (and because the tool is a linter, that file configures lints).

---

The reason I think a `[lints]` section makes sense separated from a `[rules.<tool>]` section (or similar) is, that lints are a first class compiler feature. Tools are more a `cargo` feature[^1].

[^1]: And probably the tools will only pick up those configurations when specified in the `Cargo.toml` when run through their cargo subcommand. At least this would be the case for Clippy and IIUC also for `rustfmt`.

Then the extension of the `[lints]` section with a `[lints.<tool>]` section is analogue to tool lints, which are also a first class compiler feature.

---

So I would not classify tool configuration as a future possibility/extension of this PR[^2], but as a complete different discussion. We might or might not want to include a `metadata` section for tools in the future, but this is IMO out of the scope of this RFC. This RFC should focus solely on if and how lint **levels** can be set in the `Cargo.toml` - for builtin compiler lints and for tool lints.

[^2]: I know that in earlier threads I stated something different, but reading through the discussions and thinking about it more I changed my mind.

Reviewer 7: > My revised thinking is that this RFC should consider changing from [lints] to [rules] and keep everything else the same. I think [rules] is just enough additional generality over [lints] because "lint" usually means the effect is just that some warning message(s) will be generated at some point, whereas a "rule" may actually be acted upon in some general way, for example rustfmt formatting a file.

This looks like a strong argument to keep them separate. They have very different behaviour and use cases for the end user. For example, an IDE would want to read the `[lints]` section and extract projectwide lint configuration for its inline warnings. If lints were just some `[rules.clippy.<lint>]` values, this could cause confusion between lints and arbitrary clippy options and would require some complex disambiguation on the IDE side. As a user, it's also very desirable to be able to tell at a glance whether the project changes lint settings in `Cargo.toml` and in what way. Finally, since lints don't change the compiler output, they could be ignored for the purposes of artifact caching. For general tooling rules this assumption may be false.

Reviewer 10: Sorry for not coming back to this for so long, but at this point I believe that the points are much stronger that this RFC should stick with `[lints]` as-is and not try to generalize any further. I appreciate the consideration and thought put into the discussion here!

lints from `lint = level` to `--level lint`, sort them by priority and then
lint name, and pass them on the command line before other configuration,
`RUSTFLAGS`, allowing user configuration to override package configuration.
These flags will be fingerprinted so changing them will cause a rebuild only
for the commands where they are used.

Initially, the only supported tools will be:
- `rust`
- `clippy`
- `rustdoc`

Addition of third-party tools would fall under their
[attributes for tools](https://github.com/rust-lang/rust/issues/44690).

**Note:** This reserves the tool name `workspace` to allow workspace inheritance.

# Drawbacks
[drawbacks]: #drawbacks

There has been some user/IDE confusion about running commands like `rustfmt`
directly and expecting them to pick up configuration only associated with their
higher-level cargo-plugins despite that configuration (like `package.edition`)
being cargo-specific.  By baking the configuration for rustc, rustdoc, and
clippy directly into cargo, we will be seeing more of this.  A hope is that
this will actually improve with this RFC.  Over time, tools will need to switch

> to the model of running `cargo` to get confuguratio in response to this RFC.

Reviewer 14: Typo?

```suggestion
to the model of running `cargo` to get configuration in response to this RFC.
```

- [EmbarkStudios/rust-ecosystem#59](https://github.com/EmbarkStudios/rust-ecosystem/issues/59)
- [Proposal: Cargo Lint configuration](https://internals.rust-lang.org/t/proposal-cargo-lint-configuration/9135)

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

A new `lints` table would be added to configure lints:
```toml
[lints.rust]
unsafe = "forbid"
```
and `cargo` would pass these along as flags to `rustc` and `clippy`.

This would work with
[RFC 2906 `workspace-deduplicate`](https://rust-lang.github.io/rfcs/2906-cargo-workspace-deduplicate.html?highlight=2906#):
```toml
[lints]
workspace = true

[workspace.lints]
unsafe = "forbid"
```

## Documentation Updates

## The `lints` section

*as a new ["Manifest Format" entry](https://doc.rust-lang.org/cargo/reference/manifest.html#the-manifest-format)*

Override the default level of lints from different tools by assigning them to a new level in a
table, for example:
```toml
[lints.rust]
unsafe = "forbid"
```

This is short-hand for:
```toml
[lints.rust]
unsafe = { level = "forbid", priority = 0 }
```

`level` corresponds to the lint levels in `rustc`:
- `forbid`
- `deny`
- `warn`
- `allow`

> `priority` is a signed value that controls which lints or lint groups override other lint groups:

Reviewer 15: ```suggestion
`priority` is a signed number that controls which lints or lint groups override other lint groups:
```

I also wonder, should it be a float or is restricting it to integers enough?

Author: I doubt we need the flexibility of floats.  I've made it explicit about being integers.

- lower (particularly negative) numbers have lower priority, being overridden

>   by higher numbers, and shows up first on the command-line to tools like

Reviewer 15: ```suggestion
  by higher numbers, and show up first on the command-line to tools like
```

higher-level cargo-plugins despite that configuration (like `package.edition`)
being cargo-specific.  By baking the configuration for rustc, rustdoc, and
clippy directly into cargo, we will be seeing more of this.  A hope is that
this will actually improve with this RFC.  Over time, tools will need to switch
to the model of running `cargo` to get configuration in response to this RFC.
As for users, if a tool's primary configuration is in `Cargo.toml`, that will
provide a strong coupling with `cargo` in users minds as compared to using an
external configuration file and overlooking the one or two fields read from
`Cargo.toml`.

As this focuses on lints, this leaves out first-party tools that need
configuration but aren't linters, namely `rustfmt`, leading to an inconsistent
experience if `clippy.toml` goes away.

A concern brought up in
[rust-lang/rust-clippy#1313](https://github.com/rust-lang/rust-clippy/issues/1313)
was that this will pass lints unconditionally to the underlying tool, leading
to "undefined lint" warnings when used on earlier versions, requiring that
warning to also be suppressed, reducing its value.  However, in the "Future

> possibility's", we mention direct support for tying lints to rust versions.

Reviewer 15: ```suggestion
possibilites" section, we mention direct support for tying lints to rust versions.
```

  workspace

Instead of the `[lints]` table being `lint = "level"`, we could organize
it around `level = ["lint", ...]` like some other linters do (like
[ruff](https://beta.ruff.rs/docs/configuration/)) but this works better for
logically organizing lints, highlighting what changed in diffs, and for
possibly adding lint-specific configuration in the future.

## Workspace Inheritance

Instead of using workspace inheritance for `[lint]`, we could make it
workspace-level configuration, like `[patch]` which is automatically applied to
all workspace members.  However, `[patch]` and friends are because they affect
the resolver / `Cargo.toml` and so they can only operate at the workspace
level.  `[lints]` is more like `[dependencies]` in being something that applies
at the package level but we want shared across workspaces.

Instead of traditional workspace inheritance where there is a single value to
inherit with `workspace = true`, we could have `[workspace.lints.<preset>]`
which defines presets and the user could do `lints.<preset> = true`.  The user
could then name them as they wish to avoid collision with rustc lints.

> ## Lint Predence

Reviewer 15: ```suggestion
## Lint Precedence
```

workspace-level configuration, like `[patch]` which is automatically applied to
all workspace members.  However, `[patch]` and friends are because they affect
the resolver / `Cargo.toml` and so they can only operate at the workspace
level.  `[lints]` is more like `[dependencies]` in being something that applies
at the package level but we want shared across workspaces.

Instead of traditional workspace inheritance where there is a single value to
inherit with `workspace = true`, we could have `[workspace.lints.<preset>]`
which defines presets and the user could do `lints.<preset> = true`.  The user
could then name them as they wish to avoid collision with rustc lints.

## Lint Precedence

The priority field is meant to allow mimicking
- `-Aclippy::all -Wclippy::doc_markdown`
- `-D future-incompatible -A semicolon_in_expressions_from_macros`

We can't order lints based on the level as which we want first is dependent on the context.

We can't rely on the order of the keys in the table as that is undefined in TOML.

We could use an array instead of a table:

Unconfigurable:
```toml
[lints]
clippy = [
  { all = "allow" },

>   { doc_markdown = "worn" },

Reviewer 7: ```suggestion
  { doc_markdown = "warn" },
```

]
```

Configurable:
```toml
[[lints.clippy.all]]
level = "allow"
[[lints.clippy.doc_markdown]]

> level = "worn"

Reviewer 7: ```suggestion
level = "warn"
```

  - This puts the burden on the caller
- Through `RUSTFLAGS`, like `RUSTFLAGS=--forbid=unsafe cargo clippy`
  - This puts the burden on the caller
- In `.cargo/config.toml`'s `target.*.rustflags`
  - This couples you to the running in specific directories and not running in
    the right directory causes rebuilds
  - The cargo team has previously stated that
    [they would like to see package-specific config moved to manifests](https://internals.rust-lang.org/t/proposal-move-some-cargo-config-settings-to-cargo-toml/13336/14?u=epage)

We would like a solution that makes it easier to share across targets and
packages for all tools.

See also
- [rust-lang/rust-clippy#1313](https://github.com/rust-lang/rust-clippy/issues/1313)
- [rust-lang/cargo#5034](https://github.com/rust-lang/cargo/issues/5034)
- [EmbarkStudios/rust-ecosystem#59](https://github.com/EmbarkStudios/rust-ecosystem/issues/59)
- [Proposal: Cargo Lint configuration](https://internals.rust-lang.org/t/proposal-cargo-lint-configuration/9135)

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

A new `lints` table would be added to configure lints:
```toml

> [lints.rust]

Reviewer 5: I think it complicates the model quite a bit, though: now if you have `lints.x.y` it's ambiguous between "tool `x`, lint `y`" and "lint `x`, level `y`" or whatever

> [lints.rust]

Reviewer 2: Ah right. Then this has to be namespaced.

> [lints.rust]

Author: In a quick scan, I didn't see this called out in the RFC (it has come up in the thread somewhere) so I added de440582

> unsafe = "forbid"

Author: This is covered in the alternatives section

> Instead of the `[lints]` table being `lint = "level"`, we could organize
it around `level = ["lint", ...]` like some other linters do (like
[ruff](https://beta.ruff.rs/docs/configuration/)) but this works better for
logically organizing lints, highlighting what changed in diffs, and for
possibly adding lint-specific configuration in the future.

> Also other tools (like cargo-cranky) already use this syntax, and going through the cargo issue about this feature the level = ["lint1", "lint2", ...] is what most people expect it to be.

I explored this a bit in the prior art section.  Python tends to use `level = [lint]`.  golangci-lint does the same,.  Javascript's eslint and Ruby's rubocop are closer to what this RFC references.   Which people expect is likely related to what ecosystem they come from.  Even that isn't always a sure bet.  Most of my exposure to linters is with Python and the `level = [lint]` model always bothered me.

> unsafe = "forbid"

Reviewer 2: > Most of my exposure to linters is with Python and the `level = [lint]` model always bothered me.

Most of my exposure with lint configuration is also with `python`, and I come to like the `level = [lint]` style, because I find it easier to write and read a list of lints under a category, rather then go through each lint to know which one is allowed and which one is denied.

Why does this way of writing it bothers you?

I guess this is really highly dependent on personal preference...

> unsafe = "forbid"

Reviewer 16: I also think that this is more egonomic:
```text
[lints]
<level>=[
  "<rules>",
  "<rules>",
  ...
]
```

concrete example:
```toml
[lints]
allow = [
  "dead_code", 
  "unused",
  "clippy::module_name_repetitions",
]
warn = [
  "clippy::disallowed-names",
]
```

> unsafe = "forbid"

Author: One input (not requirement) for the lint format is whether we leave the door open for lint configuration (as mentioned in that block I quoted above).  At least to me, lint configuration with the `level = lint` organization is more awkward.

> unsafe = "forbid"

Reviewer 2: I would say that lint configuration is an edge case. And the only application is `priority` or potentially `rust-version`. This could also be modeled with

```toml
[lints.rust]
allow = [
  { name="dead_code", priority=1 }, 
]
```

instead of 

```toml
[lints.rust]
dead_code = { level="allow", priority=1 }
```

> unsafe = "forbid"

Author: We discussed this in the cargo team meeting.  While I might not be capturing all the views of the attendees, the thoughts were:
- One element of this conversation is what people are used to
- For `level = [lint]`
  - Less verbose
- For `lint = level`
  - Some felt that they were more likely to think "lint" and look that up, rather than "level" then "lint".
  - You don't get lost in [large lint lists](https://github.com/EmbarkStudios/rust-ecosystem/blob/main/lints.toml) like people do in dependency tables today
  - Depending on the users mental model, if we do add overriding of inherited lint configuration, the package's table being merged into the workspace table conceptually makes more sense (and is easier to implement correctly)
  - The potential for lint configuration as mentioned

Overall, we are leaning towards `lint = level` but the priority conversation could change that.

> unsafe = "forbid"

Reviewer 13: > You don't get lost in [large lint lists](https://github.com/EmbarkStudios/rust-ecosystem/blob/main/lints.toml) like people do in dependency tables today

IMO `level = [lint]` actually fares better in this sort of test, `lint = level` has two major downsides to scannability: the jagged ` = "warn"` edge makes picking out the lint names harder, and the ability to mix levels makes it easy to mis-identify what level a specific lint is at.

```
warn = [
  "clippy::all",
  "clippy::await_holding_lock",
  "clippy::char_lit_as_u8",
  "clippy::checked_conversions",
  "rust_2018_idioms",
]
deny = [
  "unsafe_code",
]
```
```
"clippy::all" = "warn"
"clippy::await_holding_lock" = "warn"
"clippy::char_lit_as_u8" = "warn"
"clippy::checked_conversions" = "warn"
"unsafe_code" = "deny"
"rust_2018_idioms" = "warn"
```

> unsafe = "forbid"

Reviewer 2: > * Some felt that they were more likely to think "lint" and look that up, rather than "level" then "lint".

I'm not sure, I understand what this is supposed to mean.

> * You don't get lost in [large lint lists](https://github.com/EmbarkStudios/rust-ecosystem/blob/main/lints.toml) like people do in dependency tables today

I also don't understand this point and think it's the other way around. I don't see how one would get less lost with the `lint = "level"` syntax:

<details>
<summary>Linked example with both syntaxes</summary>

```toml
[lints.rust]
unsafe_code = "deny"
future_incompatible = "warn"
nonstandard_style = "warn"
rust_2018_idioms = "warn"

[lints.clippy]
all = "warn"
await_holding_lock = "warn"
char_lit_as_u8 = "warn"
checked_conversions = "warn"
dbg_macro = "warn"
debug_assert_with_mut_call = "warn"
doc_markdown = "warn"
empty_enum = "warn"
enum_glob_use = "warn"
exit = "warn"
expl_impl_clone_on_copy = "warn"
explicit_deref_methods = "warn"
explicit_into_iter_loop = "warn"
fallible_impl_from = "warn"
filter_map_next = "warn"
flat_map_option = "warn"
float_cmp_const = "warn"
fn_params_excessive_bools = "warn"
from_iter_instead_of_collect = "warn"
if_let_mutex = "warn"
implicit_clone = "warn"
imprecise_flops = "warn"
inefficient_to_string = "warn"
invalid_upcast_comparisons = "warn"
large_digit_groups = "warn"
large_stack_arrays = "deny"
large_types_passed_by_value = "warn"
let_unit_value = "warn"
linkedlist = "warn"
lossy_float_literal = "warn"
macro_use_imports = "warn"
manual_ok_or = "warn"
map_err_ignore = "warn"
map_flatten = "warn"
map_unwrap_or = "warn"
match_on_vec_items = "warn"
match_same_arms = "warn"
match_wild_err_arm = "warn"
match_wildcard_for_single_variants = "warn"
mem_forget = "warn"
mismatched_target_os = "warn"
missing_enforced_import_renames = "warn"
mut_mut = "warn"
mutex_integer = "warn"
needless_borrow = "warn"
needless_continue = "warn"
needless_for_each = "warn"
option_option = "warn"
path_buf_push_overwrite = "warn"
ptr_as_ptr = "warn"
rc_mutex = "warn"
ref_option_ref = "warn"
rest_pat_in_fully_bound_structs = "warn"
same_functions_in_if_condition = "warn"
semicolon_if_nothing_returned = "warn"
single_match_else = "warn"
string_add_assign = "warn"
string_add = "warn"
string_lit_as_bytes = "warn"
string_to_string = "warn"
todo = "warn"
trait_duplication_in_bounds = "warn"
unimplemented = "warn"
unnested_or_patterns = "warn"
unused_self = "warn"
useless_transmute = "warn"
verbose_file_reads = "warn"
zero_sized_map_values = "warn"
```

As an exercise for the reader, I've hidden a "deny" in the list. Can you spot it? 

---

```toml
[lints.rust]
deny = [
    "unsafe_code",
]
warn = [
    "future_incompatible",
    "nonstandard_style",
    "rust_2018_idioms",
]

[lints.clippy]
warn = [
    "all",
    "await_holding_lock",
    "char_lit_as_u8",
    "checked_conversions",
    "dbg_macro",
    "debug_assert_with_mut_call",
    "doc_markdown",
    "empty_enum",
    "enum_glob_use",
    "exit",
    "expl_impl_clone_on_copy",
    "explicit_deref_methods",
    "explicit_into_iter_loop",
    "fallible_impl_from",
    "filter_map_next",
    "flat_map_option",
    "float_cmp_const",
    "fn_params_excessive_bools",
    "from_iter_instead_of_collect",
    "if_let_mutex",
    "implicit_clone",
    "imprecise_flops",
    "inefficient_to_string",
    "invalid_upcast_comparisons",
    "large_digit_groups",
    "large_stack_arrays",
    "large_types_passed_by_value",
    "let_unit_value",
    "linkedlist",
    "lossy_float_literal",
    "macro_use_imports",
    "manual_ok_or",
    "map_err_ignore",
    "map_flatten",
    "map_unwrap_or",
    "match_on_vec_items",
    "match_same_arms",
    "match_wild_err_arm",
    "match_wildcard_for_single_variants",
    "mem_forget",
    "mismatched_target_os",
    "missing_enforced_import_renames",
    "mut_mut",
    "mutex_integer",
    "needless_borrow",
    "needless_continue",
    "needless_for_each",
    "option_option",
    "path_buf_push_overwrite",
    "ptr_as_ptr",
    "rc_mutex",
    "ref_option_ref",
    "rest_pat_in_fully_bound_structs",
    "same_functions_in_if_condition",
    "semicolon_if_nothing_returned",
    "single_match_else",
    "string_add_assign",
    "string_add",
    "string_lit_as_bytes",
    "string_to_string",
    "todo",
    "trait_duplication_in_bounds",
    "unimplemented",
    "unnested_or_patterns",
    "unused_self",
    "useless_transmute",
    "verbose_file_reads",
    "zero_sized_map_values",
]
```
</details>

> * Depending on the users mental model, if we do add overriding of inherited lint configuration, the package's table being merged into the workspace table conceptually makes more sense (and is easier to implement correctly)

I wouldn't put a lot of thought behind what's easier to implement in `cargo`, but rather what's easier to use.

Merging `warn` tables vs merging tables of lint are not two very different mental models and both are equally easy to understand IMO.

> * The potential for lint configuration as mentioned

That is still possible by allowing tables in the `warn`/`allow`/... lists.

---

Another argument for the `level = [lint]` syntax is, that it potentially doesn't require splitting up the lint tables by tool. See https://github.com/rust-lang/rfcs/pull/3389#discussion_r1153145977. But this might still be necessary, because `rustdoc` is an outlier  IIUC.

> unsafe = "forbid"

Reviewer 13: AFAIK `rustdoc` isn't an outlier, it also ignores all unknown tool lints (`cargo rustdoc -- -W clippy::pedantic` runs just fine).

> unsafe = "forbid"

Author: > > You don't get lost in [large lint lists](https://github.com/EmbarkStudios/rust-ecosystem/blob/main/lints.toml) like people do in dependency tables today
>
> IMO level = [lint] actually fares better in this sort of test, lint = level has two major downsides to scannability: the jagged  = "warn" edge makes picking out the lint names harder, and the ability to mix levels makes it easy to mis-identify what level a specific lint is at.

When looking at a specific lint, you know what its level is vs being in a long list of lints and losing track of what level you are in.

What you are mentioning is something different, quickly scanning for lint levels.

> I wouldn't put a lot of thought behind what's easier to implement in cargo, but rather what's easier to use.

To be clear, I wasn't.  I emphasized the user workflow and parenthetically talked about the implementation.

> > The potential for lint configuration as mentioned
>
> That is still possible by allowing tables in the warn/allow/... lists. See https://github.com/rust-lang/rfcs/pull/3389#discussion_r1155824507

All things being (mostly) equal, we aren't too big of fans of complex/deep nested inline data structures.  We only considered this syntax in an alternative to `priority` because our concerns for `priority` helped shift the balance enough.

> unsafe = "forbid"

Reviewer 2: > When looking at a specific lint, you know what its level is vs being in a long list of lints and losing track of what level you are in.

I would expect the question "What lints are set to warn/deny?" come up more often, than "What level does this exact lint have?". And the second question is also easily answerable by just jumping to the start of the list with the `level = [lint]` syntax. The first question is really hard to answer with the `lint = level` syntax, if the lints are not sorted by level (which the other syntax does).

> To be clear, I wasn't. I emphasized the user workflow

ACK. I still don't see a benefit of either syntax for the user regarding packages/workspaces.

> All things being (mostly) equal, we aren't too big of fans of complex/deep nested inline data structures

I agree that the less nesting of the `lint = level` syntax is nicer for this use case. However, I don't think this is the main use case - far from it. Most lints won't need any configuration or `priority`. So I don't think this outweighs the benefits of the `level = [lint]` syntax.

> unsafe = "forbid"

Author: I've updated the RFC to include some of the discussion from here on which schema to use

```

> and `cargo` would pass these along as flags to `rustc` and `clippy`.

Author: Thanks!  This is in a54d985c

This would work with
[RFC 2906 `workspace-deduplicate`](https://rust-lang.github.io/rfcs/2906-cargo-workspace-deduplicate.html?highlight=2906#):
```toml
[lints]
workspace = true

[workspace.lints]
unsafe = "forbid"
```

## Documentation Updates

## The `lints` section

*as a new ["Manifest Format" entry](https://doc.rust-lang.org/cargo/reference/manifest.html#the-manifest-format)*

Override the default level of lints from different tools by assigning them to a new level in a
table, for example:
```toml
[lints.rust]
unsafe = "forbid"
```

This is short-hand for:
```toml
[lints.rust]
unsafe = { level = "forbid", priority = 0 }
```

`level` corresponds to the lint levels in `rustc`:
- `forbid`
- `deny`
- `warn`
- `allow`

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Reviewer 5: Yeah I suggested this in https://github.com/rust-lang/rfcs/pull/3389#discussion_r1107386273 as well

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Reviewer 5: I think the "in the order they are specified" model is _useful_ for when lint levels are being specified in multiple places. But I think it might be fine to do it just for CLI flags.

Strawman proposal: `rustc` gains a `sort-lint-cli-flags` arg that Cargo passes in, which perhaps becomes the default mode in the next edition.

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Reviewer 13: Does the `warnings` meta-group complicate this at all? (AFAICT with some quick testing of `-Dwarnings` behavior it ignores flag order).

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Reviewer 5: It's not actually a lint group; it turns _all_ warnings into errors, including non-lint warnings

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Reviewer 2: > Strawman proposal: `rustc` gains a `sort-lint-cli-flags` arg that Cargo passes in, which perhaps becomes the default mode in the next edition.

I like that better, than adding completely new, but almost identical flags for this.

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Author: @GilShoshan94 I would ask that we keep the conversation more focused.  There is another recent thread on `level = [lint]`.  As for not using sub-tables, please open a new thread for that.  This is focused on `priority` and it will make things complicated to cover other topics.

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Reviewer 2: @GilShoshan94, it is not about priority of the source where the lint is allowed, but about priority in the `Cargo.toml` file itself. E.g.

```toml
[lints.clippy]
allow = ["pedantic"]
warn = ["bool_to_int_with_if"]
```

should it be passed to `rustc`/`clippy` as `-Aclippy::pedantic -Wclippy::bool_to_int_with_if` or `-Wclippy::bool_to_int_with_if -Aclippy::pedantic`. In the first case `bool_to_int_with_if` would warn, in the second case it would not. In this case it is obvious that the user intend is the former, but cargo doesn't know that `pedantic` is a group and overwrites `bool_to_int_with_if`. This gets even more complicated when multiple (overlapping) lint groups are involved.

The question is, can we come up with a heuristic that feels natural to the user or do we need a `pedantic` configuration?

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Reviewer 17: I'd expect a certain level of specificity here.

1. Order by location.  The latest one in the series takes precedence.
  1. Read from ~/.cargo/config.toml (if exists)
  2. Read from ./.cargo/config.toml (if exists)
  3. Read from workspace:Cargo.toml (if exists)
  4. Read from crate:Cargo.toml
  5. Read from inlined attributes
2. Within each location, override by specificity. `-Aclippy::pedantic` is less specific than `-Wclippy::bool_to_int_with_if`.  Therefore, irrespective of order, `-Wclippy::bool_to_int_with_if` should be at WARN, and the rest of pedantic should be at ALLOW.

If cargo doesn't know that pedantic is a group, it shouldn't be that hard to add that as a notation and order as such.

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Author: @spmadden auto-sorting the lint groups is the proposal being discussed in this thread.  One of the questions at hand with the heuristic approach is how to deal with lint groups that intersect, like `rust-2018-idioms` and `rust-2021-compatibility`.

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Author: Auto-sorting and heuristics do not have to be mutually exclusive.  We can start with one now and add the other later.  The basic idea is that until auto-sorting is added, the order within a priority is unspecified.  Once auto-sorting is added,we'll auto-sort within a priority level.  For this to work, we'd need
- the schema to support adding `priority` (likely `lint = level`)
- the CLI flag to accept a group of lints w/ levels to sort so we can pass it for each priority level
  - I suspect this would already be needed so that we don't auto-sort between lint level sources, keeping the existing priority system working for those sources

I lean towards starting with priority because its a lower level, lower policy mechanism allowing this to be get in sooner and we can iterate on policy ideas for auto-sorting over time in nightly.

Granted, there might be personal biases in there.
- I'm unsure if its originally my idea but I put it in the RFC (was originally proposed in a cargo team meeting I think)
- Its a simpler,. easy to implement, and isolated change rather than to work across rustc, clippy, and rustdoc to get buy-in on the flag and the heuristics, then implement it across all the tools, and then get the enough testing on the heuristics to be satisfied
- I'm familiar with cargo so I can more quickly get the needed changes in rather than having to dig into unfamiliar code or hope someone else picks it up

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Reviewer 2: > we can iterate on policy ideas for auto-sorting over time in nightly.

I think this is a good idea. Let's go with the `priority` attribute first and while this is in nightly we can try to find a auto-sorting policy. If we manage to come up with an auto-sorting policy that gets rid of the `priority` attribute, then that would be the best case.

So I recommend to resolve this, by adding a note about the `priority` attribute that it is subject to removal, before the feature gets stabilized and then reference a Future Possibility about an auto-sorting feature.

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Reviewer 2: >  My hope is to stablize this sooner than later; to decouple the auto-sort policy work from getting this into people's hands on stable.

My concern is, that if the `priority` attribute gets stabilized with this feature, we can't get rid of it anymore, even when we figure out a good auto-sorting policy. This might put long-term maintenance burden on the `cargo` team. But I don't know much about the cargo code base or how much hassle that is. And I won't have to maintain it, so I'll leave that decision for you.

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Author: We talked about this in the cargo team meeting. 

We are concerned locking down future decisions based on current policy and implementations.  Examples
- Could clippy change its mind about groups being disjoint?
- We haven't heard from rustc diagnostic or rustdoc folks yet on this (I'll be reaching out once I'm done writing this up)
- We have two likely new tools on the horizon (`cargo`, `cargo-semver-check`) and we don't want to make decisions for them when they haven't yet explored their problem space

Of the options listed below, I feel we are leaning towards
- Priority now, with the possibility of a mixture of auto-sorting and warnings in the future
- The RUSTFLAGS-like schema of `[ {level=lint} ]`

The plan for moving this forward is
- See where this conversation goes
- Get more information from rustc/rustdoc folks
- Ping some specific individuals who have dealt with some of these problems to make sure they have the opportunity to provide feedback

## Options

### Priority only

🔴 We are concerned with a users ability to get it right on the first try or debug it if they didn't.

### Priority with a future possibility for warning when the order doesn't make sense 

**Variant:** Instead of just warn, auto-sort within a priority level (most-specific last)

Example warnings:
- groups being in the same priority as lints within their grup
- intersecting groups in the same priority)

Why warning and not error:
- Only being a warning means previously disjoint groups can be changed to intersect without being a breaking change
- Only being a warning means we can add it later without being a breaking change

We talked about two ways for detecting warning scenarios
- Add new CLI contract for batching the lints by priority level so the tool can make the relevant decisions
- Cargo queries the linter for the lint-group graph and caches it
  - More complexity on cargo
  - Consistent experience that the cargo team can evolve,. rather than each linter implementing it

We can use the time while only priority is stable to double-check our assumption of the user confusion before committing to the future possibility

This allows us to have a fact path to a stable MVP to use

See above for a lot of the details

✅ While `priority` isn't a great experience, holistically this might be acceptable

### Auto-sort with a heuristic

🔴 We are concerned about the fact that this effectively discourages intersecting groups and would prefer the above option as it has the user break ties rather than trying to guess on behalf of the user on how to break ties.

### Use a schema more akin to RUSTFLAGS

Self-reflection: for any of these variants, are we falling into the trap of thinking something is "right" because we started with the CLI / rust-attribute solution (path-dependence)?

**Variant 1: Array of Tables**
```toml
[lints]
# inline
rust = [
    { lint = "all", level = "warn" },
    { lint = "allow_dead_code", level = "allow" },
]
# standard
[[lints.clippy]]
lint = "all"
level = "warn"
[[lints.clippy]]
lint = "cyclomatic_complexity"
level = "allow"
```
- 🔴 We are not thrilled with the syntax for this

**Variant 2: Allow a DSL shorthand**
```toml
[lints]
rust = [
    # short-flag prefix
    "Wall",
    "Aallow_dead_code",
    # more explicit
    "A=allow_dead_code",
    # Mirror the CLI
    "-Aallow_dead_code",
    "--warning=cyclomatic_complexity",
    # with lint config
    { lint = "cyclomatic_complexity", level = "warn", threshold = 50 }
]
```
- 🔴 The short-letter prefix isn't too clear
  - People are used to `Wall` but `Aallow_dead_code` is harder to parse
  - Even with people used to `-Wall`, its become a term of its own to the point that some people do not realize its short for `-W all` or `--warning all`
  - Generally short-flags are meant for humans, long flags are encouraged automation
- 🔴 Not thrilled with any of these as a DSL

**Variant 3: Mirror RUSTFLAGS in TOML**
```toml
[lints]
rust = [
    {allow="all"},
    {warn="allow_dead_code"},
]

clippy = [
    {warn="pedantic"}
]
```
- ✅ This has the feel of RUSTFLAGS without a DSL
- `level=lint` keeps the keys finite so we can add lint configuration in the future

### User-declared groups

```toml
[lints.clippy.groups]
pedantic = "warn"

[lints.clippy.lints]
unused_self = "allow"
```
- Effectively `groups` and `lints` are priority level keywords with everything in `groups` going before `lints`
- 🔴 `all` would be hard coded to go before all other groups
- 🔴 no way to solve intersecting groups
- 🔴 limits group hierarchies
- 🔴 Unless we also implement the same mechanism used for the Warning option, users could put things under the wrong priority level keyword
  - Is it big enough of a deal for us to do anything about?

*phew* I think thats it?

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Reviewer 8: Is deserializing `lints.tool` as `IndexMap<LintName, Level>` an option? As in, it's just an object with name to level mapping, but the order inside the TOML determines what overrides?

That is how many CLIs work: if you pass `--no-foo --foo`, `foo` is on, for `--foo, --no-foo` it's off.

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Reviewer 2: > * Could clippy change its mind about groups being disjoint?

Highly unlikely, but not impossible. Also other tools (rustc) already do this or will do this.

---

1. Agreed

2. I'm a bit confused: If we can detect conflicts in priority groups and warn about them, shouldn't it then be also possible to resolve those conflicts? Is resolving this always possible automagically? I think having `priority` in the MVP will give us time to figure those things out

3. Agreed

4. I hate everything about this :sweat_smile: : This is just not useable. It's all the bad parts from the `lint = "level"` and the `level = [lint]` syntaxes combined IMO.

5. is maybe an option, but if we punt this to the user, I think `priority` is the better approach.

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Author: > Is using order for priority an option¹? As in, it's just an object with name to level mapping, but the order inside the TOML determines what overrides.
> 
> That is how many CLIs work: if you pass `--no-foo --foo`, `foo` is on, for `--foo, --no-foo` it's off.
> 
> ¹ easy to maintain in deserialization w/ `IndexMap`

By making the reference to `IndexMap`, I assume you are meaning the order within a table?  The cargo team ruled that out because (1) order for keys is not defined in the TOML spec and we don't want to rely on an extension of the TOML spec and (2) this doesn't communicate semantics to the user/tool and they could easily do things incorrectly, e.g. sorting the table.

Option 4 tries to instead explore ordering within TOML syntax and semantics.

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Author: > 2 I'm a bit confused: If we can detect conflicts in priority groups and warn about them, shouldn't it then be also possible to resolve those conflicts? 

We could except for when groups intersect.  Framed differently, its the auto-sort solution but punts to the user instead of using a heuristic, with an MVP that always punts to the user.

Something I forgot to include in discussing is going past the MVP is the cargo/linter contract and acknowledging the parallel work for registering custom lints.  I've not followed that too closely.  Depending on how likely that is to be completed and what options are possibilities, each of these proposals could have a limiting factor on the other.

> `priority` is a signed integer that controls which lints or lint groups override other lint groups:

Author: I've updated the RFC to reflect Option 2 (`priority` with warnings/auto-sort in the future).

- lower (particularly negative) numbers have lower priority, being overridden
  by higher numbers, and show up first on the command-line to tools like
  `rustc`

To know which table under `[lints]` a particular lint belongs under, it is the part before `::` in the lint
name.  If there isn't a `::`, then the tool is `rust`.  For example a warning

> about `unsafe` would be `lints.rust.unsafe` but a lint about

Reviewer 2: Is `rustc` treated as a known tool, just like Clippy? I think we could/should change that, which would address this concern: `rustc` would say: "Unknown tool `rustc`". After all, `rustc` lints only make sense, when compiling rustc itself or tools like Clippy. We could just add a `#![register_tool(rustc)]` to Clippy.

> about `unsafe` would be `lints.rust.unsafe` but a lint about

Author: I've at least called this out in  86932bd

> about `unsafe` would be `lints.rust.unsafe` but a lint about

Reviewer 16: I don't see the point to have seperated lints tables for rustc and clippy (or any other lint tool). I think a single table is preferable and to have the clippy:: prefix like it is today in the atrribute inlined in the code.

```toml
[lints]
allow = [
  "dead_code", 
  "unused",
  "clippy::module_name_repetitions",
]
warn = [
  "clippy::disallowed-names",
]
```

> about `unsafe` would be `lints.rust.unsafe` but a lint about

Author: Part of the driving reason for separate tables is if we use the lints as keys to avoid having to quote the keys.  Either way, cargo will still need to process the lints to know which tool to send them to (at least in rustdocs case).

As for what format to use, that is a separate thread.

`clippy::enum_glob_use` would be `lints.clippy.enum_glob_use`.

## The `lints` table

*as a new [`[workspace]` entry](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-workspace-section)*

The `workspace.lints` table is where you define lint configuration to be inherited by members of a workspace.

Specifying a workspace lint configuration is similar to package lints.

Example:

```toml
# [PROJECT_DIR]/Cargo.toml
[workspace]
members = ["crates/*"]

[workspace.lints.rust]
unsafe = "forbid"
```

```toml
# [PROJECT_DIR]/crates/bar/Cargo.toml
[package]
name = "bar"
version = "0.1.0"

[lints]
workspace = true
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

When parsing a manifest, cargo will resolve workspace inheritance for
`lints.workspace = true` as it does with basic fields, when `workspace` is
present, no other fields are allowed to be present.  This precludes having the
package override the workspace on a lint-by-lint basis.

cargo will contain a mapping of tool to underlying command (e.g. `rust` to

> `rustc`, `clippy` to `rustc` when clippy is the driver, `rustdoc` to

Reviewer 2: Yes, Clippy calls `cargo fix`, but with Clippy lints registered in the linked `rustc` driver, so that they get run and fixed by `cargo fix`.

I wouldn't say that cargo is responsible for applying the lints. Cargo can just pass all the specified lints to the `driver` that is used. `rustc` will ignore all tool lints that are not registered. Clippy ignores `-W`/... flags altogether and just forwards those to the `rustc` driver.

I'm not sure anymore if I understand the problem statement here. I think `cargo` doesn't have to filter lints by the tool it belongs to, but can just always specify all lints.

> `rustc`, `clippy` to `rustc` when clippy is the driver, `rustdoc` to

Author: > I'm not sure anymore if I understand the problem statement here. I think cargo doesn't have to filter lints by the tool it belongs to, but can just always pass all specified lints.

cargo fingerpints all build flags to know when to rebuild.  One shortcoming of `RUSTFLAGS` is that all of them are included in fingerpinting even if they are not relevant.  Can we use the extra context that cargo has with the `lints` table to reduce rebuilds when lints not relevant to the current driver are changed.

> I wouldn't say that cargo is responsible for applying the lints. Cargo can just pass all the specified lints to the driver that is used. rustc will ignore all tool lints that are not registered. Clippy ignores -W/... flags altogether and just forwards those to the rustc driver.

What is was referring to is that cargo will be responsible for reading `[lints]` and passing them to the driver.

I will say that though cargo does identify the driver and key some UX off of that (to specialize the message on how to fix),  that doesn't mean that it is reliable enough to make behavior choices off of.  I'm also unsure how registering of custom linters would affect this.

`rustdoc`).  When running the underlying command for the specified package,
cargo will transform the lints from `lint = level` to `--level lint`, sort them
by priority and then lint name, and pass them on the command line before other
configuration, `RUSTFLAGS`, allowing user configuration to override package
configuration.  These flags will be fingerprinted so changing them will cause a
rebuild only for the commands where they are used.

Note that this means that `[lints]` does not affect dependencies which is
normally not an issue due to `--cap-lints` being used for dependencies.

Initially, the only supported tools will be:
- `rust`
- `clippy`

> - `rustdoc`

Reviewer 18: I'm not sure I understand what is meant by being ignored.

If an unknown tool is passed to rustc, it gives a hard error ([E0602](https://doc.rust-lang.org/error_codes/E0602.html)). 

Rustc only ignores lint paths for tools it knows about (clippy, rustc, rustfmt, and rustdoc) or are registered (a bit complicated). 

As for my comment about namespaces, I was mostly concerned about the idea that cargo is converting `::` to `.`. Does it convert all `::`? If just the first `::`, might that cause confusion? (This is hypothetical, since AFAIK none of the current tools bother with additional `::` separators.)

As for restricting the set of tools names, I think I would also lean towards "forward it to rustc and rustdoc and let them figure it out".  Though that may further cause problems with the 1-letter difference between `rustc` and `rust`, where if a user mistakenly adds a `rustc` table, it will essentially be ignored.

> - `rustdoc`

Reviewer 2: > I'm not sure I understand what is meant by being ignored.

`rustc` only splits on the first `::` and treats the rest of the "Path" of the lint as a plain string. So it doesn't parse the whole lint name into a `Path` and treats it like module namespaces. 

> As for my comment about namespaces, I was mostly concerned about the idea that cargo is converting `::` to `.`.

Cargo doesn't convert `::` to `.`, but rather `.` to `::`. But the `.` it converts is part of the TOML syntax and not part of a string. If you define a lint name as a string `"mytool::foo::bar"`  cargo shouldn't touch it.

Addition of third-party tools would fall under their
[attributes for tools](https://github.com/rust-lang/rust/issues/44690).

**Note:** This reserves the tool name `workspace` to allow workspace inheritance.

# Drawbacks
[drawbacks]: #drawbacks

Since `[lints]` only affects the associated package, and not dependencies, it
will not work with `future-incompat` lints that are meant to be applied to
dependencies.  This may cause some user confusion.

There has been some user/IDE confusion about running commands like `rustfmt`
directly and expecting them to pick up configuration only associated with their
higher-level cargo-plugins despite that configuration (like `package.edition`)
being cargo-specific.  By baking the configuration for rustc, rustdoc, and

> clippy directly into cargo, we will be seeing more of this.  A hope is that

Author: 155776772dbcf834c93c5d752f5eae381430d831 clarifies that I meant configured lint levels, not linter configuration.  This section came at the behest of rustfmt's maintainer because they have run into user confusion over the behavior difference between running `rustfmt` directly and `cargo fmt`.

Translated to this RFC, this would be like someone running rustc directly and being confused as to why their `Cargo.toml` lint levels were not applied.  As I said, I don't expect this to be a big deal but I went ahead and called out this potential confusion.

Go
- [golangci-lint](https://golangci-lint.run/usage/configuration/)
  - Format is `level = [lint, ...]`

Ruby
- [rubocop](https://docs.rubocop.org/rubocop/1.45/configuration.html)
  - Format is `Lint: Enabled: true`

# Unresolved questions
[unresolved-questions]: #unresolved-questions

# Future possibilities
[future-possibilities]: #future-possibilities

## rustc reporting `Cargo.toml` as lint-level source

Currently Rust tells you where a lint level was enabled when it emits a lint.
`rustc` only sees that these lints are coming in from the command-line and
doesn't know about `[lints]`.
It would be nice if it could also point to Cargo.toml for this.  This could be
as simple as a `--lint-source=Cargo.toml` with rustc knowing just enough about
the `[lints]` table to process it directly.

## External file

Like with `package.license`, users might want to refer to an external file for
their lints.  This especially becomes useful for copy/pasting lints between
projects.

## Configurable lints

We can extend basic lint syntax:
```toml
[lints.clippy]
cyclomatic_complexity = "allow"
```
to support configuration, whether for cargo or the lint tool:
```toml
[lints.clippy]

> cyclomatic_complexity = { level = "allow", rust-version = "1.23.0", threshold = 30 }

Author: imo it is a future consideration as it is an input on the schema design.  We shouldn't design assuming its a for-sure thing but if things are weighed out equally, we should probably keep the schema open to supporting it.

And yes, I understand clippy lints do not have configuration.  At the time I did not know that but I still think its useful to be concrete in examples.  I've added 2f5f8738 to try to avoid people getting the wrong idea.

```
Where `rust-version` is used by cargo to determine whether to pass along this
lint and `threshold` is used by the tool.  We'd need to define how to
distinguish between reserved and unreserved field names.

Tool-wide configuration would be in in the `lints.<tool>.metadata` table and be
completely ignored by `cargo`.  For example:
```toml
[lints.clippy.metadata]
avoid-breaking-exported-api = true
```

Tools will need `cargo metadata` to report the `lints` table so they can read
it without re-implementing workspace inheritance.

## Packages overriding inherited lints

Currently, it is a hard error to mix `workspace = true` and lints.  We could
open this up in the future for the package to override lints from the
workspace.  This would not be a breaking change as we'd be changing an error
case into a working case.  We'd need to ensure we had a path forward for the

> semantics for configurable lints.

Author: 35439673 clarifies the wording that we should take the future possibility into consideration and that it isn't a hard requirement.

However, please understand that clippy is not the only consideration in this design.  As this section didn't even mention clippy,. I feel like the insistence on this is starting to step over a line.

> semantics for configurable lints.

Reviewer 2: Thanks for clarifying it!

require a new channel to communicate these lints to the tools (unless we infer
order for flags/config as well) and require them to organize their data in a
way for it to inferred.

We could use an array instead of a table:

Unconfigurable:
```toml
[lints]
clippy = [
  { all = "allow" },
  { doc_markdown = "warn" },
]
```

Configurable:
```toml
[[lints.clippy.all]]
level = "allow"
[[lints.clippy.doc_markdown]]
level = "warn"
```
Where the order is based on how to pass them on the command-line.

Complex TOML arrays tend to be less friendly to work with including the fact
that TOML 1.0 does not allow multi-line inline tables.

# Prior art
[prior-art]: #prior-art

Rust
- [cargo cranky](https://github.com/ericseppanen/cargo-cranky)

Python
- [flake8](https://flake8.pycqa.org/en/latest/user/configuration.html)
  - Format is `level = [lint, ...]`
- [pylint](https://github.com/PyCQA/pylint/blob/main/examples/pylintrc#L402)
  - Format is `level = [lint, ...]` but the [config file is a reflection of the CLI](https://pylint.pycqa.org/en/latest/user_guide/configuration/index.html)
- [ruff](https://beta.ruff.rs/docs/configuration/)

>   - Format is `level = [lint, ...]`, likely due to past precedence in ecosystem (see above)

Author: Thanks for the clarification!  I've updated it in 45766c48

possibilities" section, we mention direct support for tying lints to rust versions.

This does not allow sharing lints across workspaces.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

This could be left to `clippy.toml` but that leaves `rustc` without a solution.

`[lints]` could be `[package.lints]`, tying it to the package unlike `[patch]`
and other fields that are more workspace related.  Instead, we used
`[dependencies]` as our model.

`[lints]` could be `[lint]` but we decided to follow the precedence of `[dependencies]`.

Instead of `<tool>.<lint>`, we could use `<tool>::<lint>` (e.g.
`"clipp::enum_glob_use"` instead of `clippy.enum_glob_use`), like in the
diagnostic messages.  This would make it easier to copy/paste lint names.
However, with the schema being `<lint> = <level>`, this would require quoting
the keys rather than leaving them as bare words.  This would also cause
problems for tool-level configuration.  The first is that the lints wouldn't be
easily grouped with their config.  The second is if we use `<lint> = <level>`,
we'd be mixing tool names in with lints, making it easier to conflict and

> harder to collect all lints to forward to a tool.

Reviewer 13: It's not necessary to collect the lints per-tool, tools automatically ignore all lints that aren't in their namespace

```rust
> cargo rustdoc -- -W clippy::pedantic
 Documenting cbor-diag v0.1.11 (/home/nemo157/sources/cbor-diag-rs)
    Finished dev [unoptimized + debuginfo] target(s) in 0.43s
> cargo clippy -- -W rustdoc::foobaz
    Checking cbor-diag v0.1.11 (/home/nemo157/sources/cbor-diag-rs)
    Finished dev [unoptimized + debuginfo] target(s) in 0.76s
```

The potential exception to this could be `cargo::` lints, since that isn't a registered tool namespace, and doesn't really make sense to be as the lints won't be configured in source code.

Author: This is talking about the data being
```toml
[lints]
"clippy::pedanic" = "warn"

[lints.clippy]
rust-version = "1.60.0" 
```
which maps to
```json
{
  "lints": {
    "clippy": {
      "rust-version": "1.60.0",
    }
    "clippy](clippy::pedanic": "warn",
  }
}
```

Now, we need to collect the lints from this to pass to the tools but which keys are for lints?  We can try to use the shape of the data but that seems brittle especially if we support lint configuration.

Reviewer 13: Ah, this is assuming the future extension of lint config. I'm not sure that it makes sense to have specifically lint config, a more general tool config could be better, that would allow configuring things other than lints, and could also be used for configuring clippy's lints:

```toml
[tool.rustdoc]
show-type-layout = true
```

Having the config under a separate section would avoid this conflict.

Author: c3f932cc tries to remove some of the confusion around this.

cargo will contain a mapping of tool to underlying command (e.g. `rust` to
`rustc`, `clippy` to `rustc` when clippy is the driver, `rustdoc` to
`rustdoc`).  When running the underlying command for the specified package,
cargo will:
1. Transform the lints from `tool.lint = level` to `--level tool::lint`
  - Leaving off the `tool::` when it is `rust`
  - cargo will error if `lint` contains `::` as the first part is assumed to be
    a tool and it should be listed in that tool's table
2. Sort them by priority and then an unspecified order within priority that we may change in the [future](#future-possibilities).
  - On initial release, the sort will likely be reverse alphabetical which would cause `all` to be last, making it unlikely to do what the user wants which would raise awareness of the need for setting `priority` for all groups.
3. Pass them on the command line before other configuration like
`RUSTFLAGS`, allowing user configuration to override package configuration.
  - These flags will be fingerprinted so changing them will cause a rebuild only
    for the commands where they are used.

Note that this means that `[lints]` does not affect dependencies which is

> normally not an issue due to `--cap-lints` being used for dependencies.

Reviewer 18: Can you clarify what this means? I see some different interpretations:

1. The `[lints]` table in Package A only applies to Package A, not its dependencies.
2. The `[lints]` table in a non-local[^1] dependency does not apply to that dependency.
3. The `[lints]` table in a non-local dependency is applied to that dependency, but usually doesn't matter because non-local dependencies have `--cap-lints`.

[^1]: Where non-local is defined by cargo's [`is_local`](https://github.com/rust-lang/cargo/blob/0ad289b264f6003a663832103e2ea810e8556dbf/src/cargo/core/compiler/unit.rs#L102-L104).

The first seems reasonable to make it clear that a `[lints]` table only applies to the package where it is defined.

The second seems a little more questionable to me. For example, a user may want to see warnings in non-local dependencies (such as with `-vv`). That may be a bit of an edge case, but something feels a little uncomfortable to me with just omitting the flags.

Author: Updated it to be

> Note that this means that `[lints]` is only applied to the package where its
defined and not to its dependencies, local or not.  This avoids having to unify
`[lints]` tables across local packages.  Normally, lints for non-local
dependencies won't be shown anyways because of `--cap-lints`.  As for local
dependencies, they will likely have their own `[lints]` table, most the same
one, inherited from the workspace.

- `warn`
- `allow`

> `priority` is a signed value that controls which lints override other lints:

Reviewer 5: nit: should mention groups here, because lints don't override other lints

  - Leaving off the `tool::` when it is `rust`
  - cargo will error if `lint` contains `::` as the first part is assumed to be
    a tool and it should be listed in that tool's table
2. Sort them by priority and then an unspecified order within priority that we may change in the [future](#future-possibilities).

>   - On initial release, the sort will likely be reverse alphabetical which would cause `all` to be last, making it unlikely to do what the user wants which would raise awareness of the need for setting `priority` for all groups.

Author: Keep in mind, this specifically says the order is unspecified and that we can change it.  Discussion of `all` is about a possible implementation.

# Drawbacks
[drawbacks]: #drawbacks

There has been some user/IDE confusion about running commands like `rustfmt`
directly and expecting them to pick up configuration only associated with their
higher-level cargo-plugins despite that configuration (like `package.edition`)
being cargo-specific.  By baking the configuration for rustc, rustdoc, and
clippy directly into cargo, we will be seeing more of this.  A hope is that
this will actually improve with this RFC.  Over time, tools will need to switch

> to the model of runnign `cargo` to get confuguratio in response to this RFC.

Reviewer 19: ```suggestion
to the model of running `cargo` to get configuration in response to this RFC.
```
